tests/phpunit/TestFramework/Coverage/CoverageCheckerTest.php

```diff
@@ -0,0 +1,516 @@
+            . 'the option for the path given is "--coverage-xml=/nowhere/coverage-xml"'
+        );
+
+        $checker->checkCoverageExists();
+    }
+
+    public function test_it_passes_existence_check_if_XML_index_is_present_and_JUnit_file_is_missing_without_JUnit_report(): void
+    {
+        $checker = new CoverageChecker(
+            false,
+            false,
+            '',
+            self::$coveragePath,
+            '/invalid/path/to/junit.xml',
+            $this->testFrameworkAdapterMock
+        );
+
+        $this->testFrameworkAdapterMock->method('hasJUnitReport')->willReturn(false);
+
+        $checker->checkCoverageExists();
+
+        $this->addToAssertionCount(1);
+    }
+
+    public function test_it_does_not_pass_existence_check_if_JUnit_file_is_missing_with_JUnit_report_with_lambda_test_framework_adapter(): void
+    {
+        $checker = new CoverageChecker(
+            false,
+            false,
+            '',
+            self::$coveragePath,
+            '/invalid/path/to/junit.xml',
+            $this->testFrameworkAdapterMock
+        );
+
+        $this->testFrameworkAdapterMock->method('hasJUnitReport')->willReturn(true);
+
+        $this->expectException(CoverageNotFound::class);
+        $this->expectExceptionMessage(sprintf(
+            'Could not find the file "/invalid/path/to/junit.xml". Please ensure that the '
+            . 'JUnit coverage report has been properly generated at the right place.',
+            self::$coveragePath
+        ));
+
+        $checker->checkCoverageExists();
+    }
+
+    public function test_it_does_not_pass_existence_check_if_JUnit_file_is_missing_with_JUnit_report_with_PHPUnit_test_framework_adapter(): void
+    {
+        $phpUnitAdapterMock = $this->createMock(PhpUnitAdapter::class);
+        $phpUnitAdapterMock->method('hasJUnitReport')->willReturn(true);
+
+        $checker = new CoverageChecker(
+            false,
+            false,
+            '',
+            self::$coveragePath,
+            '/invalid/path/to/junit.xml',
+            $phpUnitAdapterMock
+        );
+
+        $this->expectException(CoverageNotFound::class);
+        $this->expectExceptionMessage(sprintf(
+            'Could not find the file "/invalid/path/to/junit.xml". Please ensure that the '
+            . 'JUnit coverage report has been properly generated at the right place. The PHPUnit '
+            . 'option for the path given is "--log-junit=%s/junit.xml"',
+            self::$coveragePath
+        ));
+
+        $checker->checkCoverageExists();
+    }
+
+    public function test_it_does_not_pass_existence_check_if_JUnit_file_is_missing_with_JUnit_report_with_Codeception_test_framework_adapter(): void
+    {
+        $checker = new CoverageChecker(
+            false,
+            false,
+            '',
+            self::$coveragePath,
+            '/invalid/path/to/junit.xml',
+            new CodeceptionAdapter()
```

Comment:
I'm a bit in a pinch here. I tried to be smart and "mock" it by adding it at the end of the file and it works but seem to be deal breaker for phpstan :/

Maybe an alternative is to just pass the "has JUnit report" and test framework adapter name instead. That would make it easier but I'm a bit concerned about testing against the adapter name which we have no control over from here.

@maks-rafalko would I be wrong to add back the adapters as dev dependencies?

Reply:
Another way would be to add an interface for each of the external adapters, and make them use it. E.g.
```
class CodeceptionAdapter implements CodeceptionAdapterInterface {}
```
Where `CodeceptionAdapterInterface` located in the root package. Also solves problem with string constants inside AdapterInstallationDecider.

Reply:
`infection/infection` should not test adapters. Codeception, PHPSpec is not different to X, Y or Z adapter that can be added in the future. It's a responsibility of adapter to test how it works with the core.

IMO, what `infection/infection` should do is only rely on `TestFrameworkAdapter` inserface, no matter what implementation we have.

If some thing is on the interface - we test it in an abstract way. If not - we don't write "hardcoded" code for particular adapter.

> @maks-rafalko would I be wrong to add back the adapters as dev dependencies?

yep, it's incorrect and will lead to cycles in dependencies

---

We already have [`getInitialTestsFailRecommendations()`](https://github.com/infection/abstract-testframework-adapter/blob/master/src/TestFrameworkAdapter.php#L72) method for each adapter which does pretty much the same - it returns custom messages.

Reply:
Otherwise put, to move forward with this PR it is imperative to update TestFrameworkAdapter to include an additional method?

I don't think this is bad or anything, but rather to clarify.

Reply:
It's not that important of a message and really depends on how infection is being used, so I don't think it's the responsibility of the adapter to know that context.

I'll just inject the test framework name and has JUnit report instead: less clean but will work. There is the risk that the test framework adapter changes of names and mess up there but hey, it's not that critical so I think it's ok